lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1115 @@
+		impl LocalState for $state {
+			fn into_negotiation_context(self) -> NegotiationContext {
+				self.0
+			}
+		}
+	};
+	(REMOTE_STATE, $state: ident, $doc: expr) => {
+		define_state!($state, NegotiationContext, $doc);
+		impl RemoteState for $state {
+			fn into_negotiation_context(self) -> NegotiationContext {
+				self.0
+			}
+		}
+	};
+	($state: ident, $inner: ident, $doc: expr) => {
+		#[doc = $doc]
+		#[derive(Debug)]
+		struct $state($inner);
+		impl State for $state {}
+	};
+}
+
+define_state!(LOCAL_STATE, LocalChange, "We have sent a message to the counterparty that has affected our negotiation state.");
+define_state!(LOCAL_STATE, LocalTxComplete, "We have sent a `tx_complete` message and are awaiting the counterparty's.");
+define_state!(REMOTE_STATE, RemoteChange, "We have received a message from the counterparty that has affected our negotiation state.");
+define_state!(REMOTE_STATE, RemoteTxComplete, "We have received a `tx_complete` message and the counterparty is awaiting ours.");
+define_state!(NegotiationComplete, Transaction, "We have exchanged consecutive `tx_complete` messages with the counterparty and the transaction negotiation is complete.");
+define_state!(NegotiationAborted, AbortReason, "The negotiation has failed and cannot be continued.");
+
+type StateTransitionResult<S> = Result<S, AbortReason>;
+
+trait StateTransition<NewState: State, TransitionData> {
+	fn transition(self, data: TransitionData) -> StateTransitionResult<NewState>;
+}
+
+macro_rules! define_state_transitions {
+	(LOCAL_STATE, [$(DATA $data: ty, TRANSITION $transition: ident),+]) => {
+		$(
+			impl<S: LocalState> StateTransition<RemoteChange, $data> for S {
+				fn transition(self, data: $data) -> StateTransitionResult<RemoteChange> {
+					let mut context = self.into_negotiation_context();
+					let _ = context.$transition(data)?;
+					Ok(RemoteChange(context))
+				}
+			}
+		 )*
+	};
+	(REMOTE_STATE, [$(DATA $data: ty, TRANSITION $transition: ident),+]) => {
+		$(
+			impl<S: RemoteState> StateTransition<LocalChange, $data> for S {
+				fn transition(self, data: $data) -> StateTransitionResult<LocalChange> {
+					let mut context = self.into_negotiation_context();
+					let _ = context.$transition(data);
+					Ok(LocalChange(context))
+				}
+			}
+		 )*
+	};
+	(TX_COMPLETE_AS_ACK, $from_state: ident, $to_state: ident) => {
+		impl StateTransition<$to_state, &msgs::TxComplete> for $from_state {
+			fn transition(self, _data: &msgs::TxComplete) -> StateTransitionResult<$to_state> {
+				Ok($to_state(self.into_negotiation_context()))
+			}
+		}
+	};
+	(TX_COMPLETE, $from_state: ident) => {
+		impl StateTransition<NegotiationComplete, &msgs::TxComplete> for $from_state {
+			fn transition(self, _data: &msgs::TxComplete) -> StateTransitionResult<NegotiationComplete> {
+				let context = self.into_negotiation_context();
+				let tx = context.build_transaction()?;
+				Ok(NegotiationComplete(tx))
+			}
+		}
+	};
```

Comment:
@dunxen

Reply:
It's not jumping out at me since we are returning a wrapped tx in one and a wrapped context in the other. I tried to go down the route of a `?` fragment on the `$to_state` but it doesn't seem to work out. Admittedly my macro-foo is lacking, though.

Reply:
I mean only use one macro rule instead of two to define both.